architecture-docs/adr-001.md

```diff
@@ -0,0 +1,85 @@
+# ADR 001: Storage Consolidation
+
+## Changelog
+* 10-12-2019: Initial Draft
+
+## Context
+Currently (not counting Tendermint's internal storage or wallets), two processes maintain their internal storage:
+
+* chain-abci: stores the node state, Merkle trie of staking states, transaction metadata (whether spent or not),
+validator tracking, etc.
+* tx-validation enclave (TVE): sealed transaction data (of valid obfuscated transactions that have outputs)
+
+The reason for having two processes is that SGX SDK compilation is different and needs Intel SGX SDK tooling,
+so for the development convenience, the transaction validation code that needs to execute in an enclave
+is isolated. The inter-process communication is over a simple REQ-REP 0MQ socket.
+
+*Problem 1: These two storage locations need to be "in sync"*:
+
+when an obfuscated transaction arrives that spends some transaction outputs, chain-abci will do a basic validation and check if they are unspent and forward it to TVE (assuming its storage contains
+sealed transaction data of respective outputs). There is currently a naive check that TVE
+stores the latest app hash provided by chain-abci; and upon a startup, chain-abci cross-checks if TVE is in sync with it. This leads to various errors and annoyances that are usually resolved by removing all storage and syncing from scratch.
```

Comment:
What is current flow when `TVE` is not in sync with `chain-abci`?

Reply:
What do you mean by "flow"? Flow how that could happen or ?

Reply:
No. How do we re-synchronize them when this happens?

Reply:
> that are usually resolved by removing all storage and syncing from scratch.

the current usual flow is: delete everything, start from scratch.

should I rephrase that in that context / problem description?

Reply:
No. I was assuming that there must be an automated way that first tries to bring `TVE` and `chain-abci` in sync and we need to delete everything only when that mechanism fails. If there is no such mechanism, I think we need to address this ASAP. Either by separating out `chain-storage` in a different process or some other way.

Reply:
There could be some better way for certain cases -- e.g. if TVE has old app hash, chain-abci could replay (enclave) transactions that occurred between that app hash and the current one.
But these are extra complexities that wouldn't be necessary if there was only one storage.